drgn/helpers/linux/slab.py

```diff
@@ -321,3 +333,66 @@ def find_containing_slab_cache(  # type: ignore  # Need positional-only argument
         return slab.slab_cache
     except AttributeError:
         return NULL(prog, "struct kmem_cache *")
+
+
+def get_slab_aliases(prog: Program) -> Dict[str, str]:
+    """
+    Return a dict mapping slab cache name to the cache it was merged with.
+
+    The SLUB subsystem can merge caches with similar settings, as described in
+    the documentation of :func:`slab_cache_is_merged()`. However, if the SLUB
+    sysfs interface is enabled, then the slab merging is exposed as symlinks
+    between directories. So it is possible to use these data structures to
+    discover whether a given slab cache has been merged, and if so, which slab
+    cache it was merged with.
+
+    This function returns a mapping from original cache name, to merged cache
+    name. You can use this mapping to discover the correct cache to lookup via
+    :func:`find_slab_cache()`. If a slab cache name does not appear as a key in
+    this mapping, then it has not been merged with any other cache, and can
+    directly be looked up. A cache name may be mapped to itself; in this case,
+    the cache was merged with others, and the final merge cache used that
+    particular name.
+
+    >>> cache_to_merged = get_slab_aliases(prog)
+    >>> cache_to_merged["avc_xperms_data"]
+    'avc_xperms_data'
+    >>> cache_to_merged["dnotify_struct"]
+    'avc_xperms_data'
+    >>> find_slab_cache(prog, "dnotify_struct") is None
+    True
+    >>> find_slab_cache(prog, "avc_xperms_data") is None
+    False
+
+    :warning: This function will only work on kernels which are built with
+      ``CONFIG_SYSFS`` enabled.
+
+    :param prog: Program to search
+    :returns: Mapping of slab cache name to final merged name
+    """
+    children = rbtree_inorder_for_each_entry(
+        "struct kernfs_node",
+        prog["slab_kset"].kobj.sd.dir.children.address_of_(),
+        "rb",
+    )
+    link_flag = prog.constant("KERNFS_LINK")
+    cache_alias_map = {}
+    alias_true_name_map = {}
```

Comment:
Do we need these two layers of maps? If we only look at symlinks and skip over everything else, we can get the "true" name directly from `target_kn`. Something like:

```python
    link_flag = prog.constant("KERNFS_LINK")
    name_map = {}
    for child in rbtree_inorder_for_each_entry(
        "struct kernfs_node",
        prog["slab_kset"].kobj.sd.dir.children.address_of_(),
        "rb",
    ):
        if child.flags & link_flag:
            cache = container_of(
                cast("struct kobject *", child.symlink.target_kn.priv),
                "struct kmem_cache",
                "kobj",
            )
            name_map[child.name.string_().decode()] = cache.name.string_().decode()
    return name_map
```

Reply:
Ha, thanks for rewriting my first attempt code into something far simpler!